File: 23. FindIndexOfFirstOccurrence.cpp
#include <iostream>
#include <string>
using namespace std;

int strStr(string haystack, string needle) {
    int n = haystack.length(), m = needle.length();
    if (m == 0) return 0;
    for (int i = 0; i + m <= n; i++) {
        if (haystack.substr(i, m) == needle) return i;
    }
    return -1;
}

int main() {
    cout << strStr("sadbutsad", "sad") << "\n";
    cout << strStr("leetcode", "leeto") << "\n";
    cout << strStr("hello", "ll") << "\n";

    return 0;
}

File: 23. FindIndexOfFirstOccurrence.java
public class FindIndexOfFirstOccurrence {

    public static int strStr(String haystack, String needle) {
        int n = haystack.length(), m = needle.length();
        if (m == 0) return 0;
        for (int i = 0; i + m <= n; i++) {
            if (haystack.regionMatches(i, needle, 0, m)) return i;
        }
        return -1;
    }

    public static void main(String[] args) {
        System.out.println(strStr("sadbutsad", "sad"));   // 0
        System.out.println(strStr("leetcode", "leeto"));  // -1
        System.out.println(strStr("hello", "ll"));        // 2
    }
}
